src/algorithms/groundstate/gradient_grassmann.jl

```diff
@@ -0,0 +1,165 @@
+    (x, fx, gx, numfg, normgradhistory) = res
+    (state, pars) = x
+    return state, pars, normgradhistory[end]
+end
+
+
+# We separate some of the internals needed for implementing GradientGrassmann into a
+# submodule, to keep the MPSKit module namespace cleaner.
+"""
+A module for functions related to treating an InfiniteMPS in left-canonical form as a bunch
+of points on Grassmann manifolds, and performing things like retractions and transports
+on these Grassmann manifolds.
+
+The module exports nothing, and all references to it should be qualified, e.g.
+`GrassmannInfiniteMPS.fg`.
+"""
+module GrassmannInfiniteMPS
+
+using ..MPSKit
+using TensorKit
+import TensorKitManifolds.Grassmann
+
+"""
+Compute the expectation value, and its gradient with respect to the tensors in the unit
+cell as tangent vectors on Grassmann manifolds.
+"""
+function fg(x)
+    (state, pars) = x
+    # The partial derivative with respect to AL, al_d, is the partial derivative with
+    # respect to AC times CR'.
+    ac_d = [ac_prime(state.AC[v], v, state, pars) for v in 1:length(state)]
+    al_d = [d*c' for (d, c) in zip(ac_d, state.CR)]
+    g = [Grassmann.project(d, a) for (d, a) in zip(al_d, state.AL)]
+    f = real(sum(expectation_value(state, pars.opp, pars)))
+    return f, g
+end
+
+"""
+Retract a left-canonical infinite MPS along Grassmann tangent `g` by distance `alpha`.
+"""
+function retract(x, g, alpha)
+    (state, pars) = x
+    yal = similar(state.AL)  # The end-point
+    h = similar(g)  # The tangent at the end-point
+    for i in 1:length(g)
+        (yal[i], h[i]) = Grassmann.retract(state.AL[i], g[i], alpha)
+    end
+    y = (InfiniteMPS(yal, leftgauged=true), pars)
+    return y, h
+end
+
+"""
+Transport a tangent vector `h` along the retraction from `x` in direction `g` by distance
+`alpha`. `xp` is the end-point of the retraction.
+"""
+function transport!(h, x, g, alpha, xp)
+    (state, pars) = x
+    for i in 1:length(state)
+        h[i] = Grassmann.transport!(h[i], state.AL[i], g[i], alpha, xp[1].AL[i])
+    end
+    return h
+end
+
+"""
+Euclidean inner product between two Grassmann tangents of an infinite MPS.
+"""
+function inner(x, g1, g2)
+    (state, pars) = x
+    tot = sum(Grassmann.inner(a, d1, d2) for (a, d1, d2) in zip(state.AL, g1, g2))
+    return 2*real(tot)
+end
+
+"""
+Scale a tangent vector by scalar `alpha`.
+"""
+scale!(g, alpha) = g .* alpha
+
+"""
+Add two tangents vectors, scaling the latter by `alpha`.
+"""
+add!(g1, g2, alpha) = g1 + g2 .* alpha
+
+"""
+Take the L2 Tikhonov regularised inverse of a matrix `m`.
+
+The regularisation parameter is the larger of `delta` (the optional argument that defaults
+to zero) and square root of machine epsilon. The inverse is done using an SVD.
+"""
+function reginv(m, delta=zero(eltype(m)))
+    delta = max(delta, sqrt(eps(real(float(one(eltype(m)))))))
+    U, S, Vdg = tsvd(m)
+    Sinv = inv(real(sqrt(S^2 + delta^2*id(domain(S)))))
```

Comment:
I think `one(S)` is slightly easier than `id(domain(S))`, and also more future robust, e.g. if `S` at some point will be a specific `DiagonalTensorMap`, or if the storage of `S` would be a `CuArray`, then `one(S)` can easily inherit these properties.

Reply:
Done.